Source/AFResult.swift

```diff
@@ -0,0 +1,247 @@
+    ///
+    /// - Parameter closure: A closure that takes the success value of this instance.
+    /// - Returns: An `AFResult` instance, unmodified.
+    @discardableResult
+    public func withValue(_ closure: (Success) throws -> Void) rethrows -> AFResult<Success> {
+        switch self {
+        case .success(let value):
+            try closure(value)
+            return .success(value)
+        case .failure(let error):
+            return .failure(error)
+        }
+    }
+
+    /// Evaluates the specified closure when the `AFResult` is a failure, passing the unwrapped error as a parameter.
+    ///
+    /// Use the `withError` function to evaluate the passed closure.
+    ///
+    /// - Parameter closure: A closure that takes the success value of this instance.
+    /// - Returns: An `AFResult` instance, unmodified.
+    @discardableResult
+    public func withError(_ closure: (Failure) throws -> Void) rethrows -> AFResult<Success> {
+        switch self {
+        case .failure(let error):
+            try closure(error)
+            return .failure(error)
+        case .success(let value):
+            return .success(value)
+        }
+    }
+
+    /// Evaluates the specified closure when the `AFResult` is a success.
+    ///
+    /// Use the `ifSuccess` function to evaluate the passed closure without modifying the `AFResult` instance.
+    ///
+    /// - Parameter closure: A `Void` closure.
+    /// - Returns: This `AFResult` instance, unmodified.
+    @discardableResult
+    public func ifSuccess(_ closure: () throws -> Void) rethrows -> AFResult<Success> {
+        switch self {
+        case .success(let value):
+            try closure()
+            return .success(value)
+        case .failure(let error):
+            return .failure(error)
+        }
+    }
+
+    /// Evaluates the specified closure when the `AFResult` is a failure.
+    ///
+    /// Use the `ifFailure` function to evaluate the passed closure without modifying the `AFResult` instance.
+    ///
+    /// - Parameter closure: A `Void` closure.
+    /// - Returns: This `AFResult` instance, unmodified.
+    @discardableResult
+    public func ifFailure(_ closure: () throws -> Void) rethrows -> AFResult<Success> {
+        switch self {
+        case .success(let value):
+            return .success(value)
+        case .failure(let error):
+            try closure()
+            return .failure(error)
+        }
+    }
+
+    /// Returns the associated value if the result is a success, `nil` otherwise.
+    public var value: Success? {
+        switch self {
+        case .success(let value):
+            return value
+        case .failure:
+            return nil
+        }
+    }
+
+    /// Returns the associated error value if the result is a failure, `nil` otherwise.
+    public var error: Error? {
+        switch self {
+        case .success:
+            return nil
+        case .failure(let error):
+            return error
+        }
+    }
```

Comment:
Hopefully in a future PR we can get rid of all these extensions across the board. At a minimum, get them moved to internal APIs.

Reply:
I've changed the access level to `internal`.

Reply:
No need for an explicit access level, just don't make the `extension` `public`.